[PHIElimination] Handle subranges in LiveInterval updates

perlfu opened the pull request:
Add subrange tracking and handling for LiveIntervals during PHI elimination.
This requires extending MachineBasicBlock::SplitCriticalEdge to also update subrange intervals.

perlfu committed:
[PHIElimination] Handle subranges in LiveInterval updates

Add handling for subrange updates in LiveInterval preservation.

Differential Revision: https://reviews.llvm.org/D158144

jayfoad commented on llvm/lib/CodeGen/PHIElimination.cpp:
```diff
@@ -136,6 +136,7 @@ INITIALIZE_PASS_END(PHIElimination, DEBUG_TYPE,
 
 void PHIElimination::getAnalysisUsage(AnalysisUsage &AU) const {
   AU.addUsedIfAvailable<LiveVariables>();
+  AU.addUsedIfAvailable<LiveIntervals>();
```
Normally you would only add this if PHIElimination can do a better job when LiveIntervals is available. If you are just saying that PHIElimination is capable of updating LiveIntervals then you should not add this.

In this case it probably won't make any difference, but in the general case, not adding unnecessary "used if available" dependencies will allow the legacy pass manager to free unneeded analyses sooner, possibly saving some memory.

perlfu replied:
It seemed to me that if we use an analysis then we should declare that usage.
But I accept your logic and it doesn't make a different in practice here.

jayfoad commented on llvm/lib/CodeGen/MachineBasicBlock.cpp:
```diff
@@ -1294,8 +1296,18 @@ MachineBasicBlock *MachineBasicBlock::SplitCriticalEdge(
       } else if (!isLiveOut && !isLastMBB) {
         LI.removeSegment(StartIndex, EndIndex);
+        for (auto &SR : LI.subranges()) {
+          if (SR.overlaps(StartIndex, EndIndex))
+            SR.removeSegment(StartIndex, EndIndex);
```
Not your fault, but it's a shame that you have to do the `overlaps` check first, because you end up looking up the segment corresponding to StartIndex twice, once in `overlaps` and once in `removeSegment`. It would make more sense to me if `removeSegment` handled the non-overlapping case as a no-op.

perlfu replied:
I have changed removeSegment to accept these.
I cannot see any obvious reason why it should not, perhaps only to avoid other coding errors?

perlfu committed:
Address reviewer feedback:
- Modify LiveRange::removeSegment so it handles intervals with no Segment
- Update code to remove overlaps() checks
- Remove addUsedIfAvailable of LiveIntervals

MatzeB commented on llvm/lib/CodeGen/PHIElimination.cpp:
```diff
@@ -400,24 +400,50 @@ void PHIElimination::LowerPHINode(MachineBasicBlock &MBB,
+        continue;
+      }
+
+      if (DestSegment->start > NewStart) {
+        // With a single PHI removed from block the index of the copy may be
+        // lower than the original PHI. Extend live range backward to cover
+        // the copy.
```
I find this comment highly confusing, and have a hard time seeing when this condition would ever be true... There was a PHI definition `DestReg` here, no? That should mean that every segment here should have been defined at `MBBStartIndex` (where PHI definitions take effect), no? (Except maybe if a subrange was undefined, but then we would have already entered the `isDead()` case above...)

perlfu replied:
I have moved and reworded this comment for clarity.

There are two details here that matter to understanding this:
* LiveIntervals of PHIs destinations start at the PHI slot index, not the block start.
* Destination copies are not inserted in the order of the PHIs they replace.

Taking the example of the attached test.
Original PHIs look like this:
```
320B      %phi0:vgpr_32 = PHI %s0a:vgpr_32, %bb.0, %s0c:vgpr_32, %bb.1
336B      %phi1:vgpr_32 = PHI %s0b:vgpr_32, %bb.0, %s0d:vgpr_32, %bb.1
```

With intervals as follows:
```
%6 [112r,208r:0) 0@112r  weight:0.000000e+00 -- s0a
%7 [128r,224r:0) 0@128r  weight:0.000000e+00 -- s0b
%8 [208r,256B:0) 0@208r  weight:0.000000e+00 -- s0c
%9 [224r,256B:0) 0@224r  weight:0.000000e+00 -- s0d
%10 [320r,384r:0) 0@320r  weight:0.000000e+00 -- phi0
%11 [336r,384r:0) 0@336r  weight:0.000000e+00 -- phi1
```

phi0 is lowered, giving the following IR:
```
336B      %phi1:vgpr_32 = PHI %s0b:vgpr_32, %bb.0, %s0d:vgpr_32, %bb.1
344B      %phi0:vgpr_32 = COPY %12:vgpr_32
```

phi1 is lowered, giving:
```
312B      %phi1:vgpr_32 = COPY %13:vgpr_32
344B      %phi0:vgpr_32 = COPY %12:vgpr_32
```

Hence COPY for phi0 has index after original PHI, but phi1 has index before original PHI.

MatzeB commented on llvm/lib/CodeGen/PHIElimination.cpp:
```diff
@@ -400,24 +400,50 @@ void PHIElimination::LowerPHINode(MachineBasicBlock &MBB,
+      } else if (DestSegment->start < NewStart) {
+        // Otherwise, remove the region from the beginning of MBB to the copy
+        // instruction from DestReg's live interval.
+        assert(DestSegment->start >= MBBStartIndex);
```
as above this should be `==`, no?

MatzeB reviewed (commented):
* Looks mostly fine
* I have some questions on one `if` condition below.
* Should probably mention in the summary that this adds subrange liveness tracking for critical edge splitting and phi elimination.

perlfu committed:
Address reviewer feedback:
- Remove duplicate comment.
- Update comments about order of PHI nodes versus COPYs.

perlfu commented on llvm/lib/CodeGen/LiveInterval.cpp:
```diff
@@ -563,13 +563,18 @@ VNInfo *LiveRange::extendInBlock(SlotIndex StartIdx, SlotIndex Kill) {
 
-/// Remove the specified segment from this range.  Note that the segment must
-/// be in a single Segment in its entirety.
+/// Remove the specified interval from this live range.
+/// Does nothing if interval is not part of this live range.
+/// Note that the interval must be within a single Segment in its entirety.
```
Done.

perlfu commented:
Ping

perlfu commented:
I will merge this on the 9th of November if there is no further feedback.

jayfoad commented on llvm/include/llvm/CodeGen/LiveIntervals.h:
```diff
@@ -139,6 +139,10 @@ class VirtRegMap;
       return LI;
     }
 
+    LiveInterval &getOrCreateEmptyInterval(Register Reg) {
```
Should really have a doxygen comment, even if it's obvious.

jayfoad reviewed (commented):
Looks good but I will defer to @MatzeB.

perlfu committed:
Add doxygen comment for getOrCreateEmptyInterval.

arsenm commented on llvm/lib/CodeGen/PHIElimination.cpp:
```diff
@@ -400,24 +400,48 @@ void PHIElimination::LowerPHINode(MachineBasicBlock &MBB,
+
+    SlotIndex NewStart = DestCopyIndex.getRegSlot();
+
+    SmallVector<LiveRange *> ToUpdate;
+    ToUpdate.push_back(&DestLI);
```
Can fold this into initializer

perlfu replied:
Done

perlfu committed:
Simplify initialisation of ToUpdate vector.
